### ISSUE TITLE
Running an RCNN on Vega 64 has the same performance as running on CPU

### ISSUE THREAD

**Issue by vidullan** [ROCm/tensorflow-upstream#343]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

ISSUE:

I am currently trying to train a model from tensorflow's model zoo (mask RCNN) and am finding that it takes 3.8 seconds per step compared to 3.3 seconds running on an i7 6700 CPU and 0.2 seconds on a nvidia 1070.

I am running this example (https://medium.com/@vijendra1125/custom-mask-rcnn-using-tensorflow-object-detection-api-101149ce0765) with different images of size (1296x864). 

In addition, the RCNN sdd model from tensorflow zoo does not even run and segfaults. 

Is this an issue with my setup or is there something very wrong with the software/hardware?

**System information**

OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 

Ubuntu 18 with kernel 4.15.0-45-generic, amdgpu, 2.0-89, 4.15.0-43-generic, x86_64: installed

TensorFlow installed from (source or binary): 
Installed using pip tensorflow-rocm

- TensorFlow version (use command below): 
tensorflow-rocm 1.12.0 

- Python version: 
Python 3.5.6 with GCC 7.3.0

- ROCm/MIOpen version:

=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (number of timestamp)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD Ryzen 7 1700 Eight-Core Processor
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0                                  
  Queue Min Size:          0                                  
  Queue Max Size:          0                                  
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768KB                            
  Chip ID:                 0                                  
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):3000                               
  BDFID:                   0                                  
  Compute Unit:            16                                 
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    16418888KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    16418888KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
  ISA Info:                
    N/A                      
*******                  
Agent 2                  
*******                  
  Name:                    gfx900                             
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128                                
  Queue Min Size:          4096                               
  Queue Max Size:          131072                             
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16KB                               
  Chip ID:                 26751                              
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):1630                               
  BDFID:                   4096                               
  Compute Unit:            64                                 
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      FALSE                              
  Wavefront Size:          64                                 
  Workgroup Max Size:      1024                               
  Workgroup Max Size Per Dimension:
    Dim[0]:                  67109888                           
    Dim[1]:                  268436480                          
    Dim[2]:                  0                                  
  Grid Max Size:           4294967295                         
  Waves Per CU:            40                                 
  Max Work-item Per CU:    2560                               
  Grid Max Size per Dimension:
    Dim[0]:                  4294967295                         
    Dim[1]:                  4294967295                         
    Dim[2]:                  4294967295                         
  Max number Of fbarriers Per Workgroup:32                                 
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    8372224KB                          
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        FALSE                              
    Pool 2                   
      Segment:                 GROUP                              
      Size:                    64KB                               
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Alignment:         0KB                                
      Acessible by all:        FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx900          
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Dimension: 
        Dim[0]:                  67109888                           
        Dim[1]:                  1024                               
        Dim[2]:                  16777217                           
      Workgroup Max Size:      1024                               
      Grid Max Dimension:      
        x                        4294967295                         
        y                        4294967295                         
        z                        4294967295                         
      Grid Max Size:           4294967295                         
      FBarrier Max Size:       32                                 
*** Done *** 

- GPU model and memory:

== cat /etc/issue ===============================================
Linux CrippledBeast 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
VERSION="18.04.1 LTS (Bionic Beaver)"
VERSION_ID="18.04"
VERSION_CODENAME=bionic

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux CrippledBeast 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy               1.14.5     
protobuf            3.6.1      
tensorflow-rocm     1.12.0     

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.12.0
tf.GIT_VERSION = v1.12.0-871-gf480b4a
tf.COMPILER_VERSION = v1.12.0-871-gf480b4a
Sanity check: array([1], dtype=int32)
WARNING:tensorflow:From /home/vidullan/miniconda3/envs/tf/lib/python3.5/site-packages/tensorflow/python/ops/distributions/distribution.py:265: ReparameterizationType.__init__ (from tensorflow.python.ops.distributions.distribution) is deprecated and will be removed after 2019-01-01.
Instructions for updating:
The TensorFlow Distributions library has moved to TensorFlow Probability (https://github.com/tensorflow/probability). You should update all references to use `tfp.distributions` instead of `tf.distributions`.
WARNING:tensorflow:From /home/vidullan/miniconda3/envs/tf/lib/python3.5/site-packages/tensorflow/python/ops/distributions/bernoulli.py:169: RegisterKL.__init__ (from tensorflow.python.ops.distributions.kullback_leibler) is deprecated and will be removed after 2019-01-01.
Instructions for updating:
The TensorFlow Distributions library has moved to TensorFlow Probability (https://github.com/tensorflow/probability). You should update all references to use `tfp.distributions` instead of `tf.distributions`.

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
./rocm.py: line 105: nvidia-smi: command not found

== cuda libs  ===================================================

**Comment by sunway513** [comment]
Hi @vidullan , to better assist with your issue, could you first try to run your applications inside our official docker image? 
With that we can further tell if the issue is related to your local environment or Tensorflow-ROCm software stack.
If you have amdgpu driver configured properly, the following steps can get you the docker environment:
```
curl -sSL https://get.docker.com/ | sh
alias drun='sudo docker run -it --network=host --device=/dev/kfd --device=/dev/dri --group-add video --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -v $HOME/dockerx:/dockerx -v /data/imagenet/caffe:/imagenet -v $HOME:/local'
drun rocm/tensorflow:rocm2.1-tf1.13-python3
```
If the issue can be reproducible inside the docker image, it'll be helpful if you can commit your docker image and provide us the exact steps to reproduce.
Some related documents:
- docker commit: https://docs.docker.com/engine/reference/commandline/commit/
- docker push: https://docs.docker.com/engine/reference/commandline/push/

**Comment by vidullan** [comment]
Hi @sunway513, unfortunately I won't be able to install a docker image. 

I tried updating my version of rocm to 2.1 and now when I run tensorflow-rocm it does not seem to be using much of the GPU at all. Even one step does not seem to complete for upwards of a minute. 

Is there a way for me to roll back to 1.9 using apt-get or do I have to build it from source?

**Comment by sunway513** [comment]
Hi @vidullan , did the application work when you had ROCm1.9? This really sounds like a local environmental issue.
You can use the following steps to clean your current ROCm packages:
`sudo apt autoremove -y rocm_bandwidth_test rocminfo hsakmt-roct hsa-rocr-dev rocm-cmake hsa-ext-rocr-dev rocm-device-libs hip_base hip_samples rock-dkms`

Then you can download the ROCm1.9.2 archieve and install the deb packages, the following steps should work:
```
cd ~ && wget http://repo.radeon.com/rocm/archive/apt_1.9.2.tar.bz2 && tar -xvf apt_1.9.2.tar.bz2
cd ~/apt_1.9.2.307 && cd pool/main && find . -mindepth 2 -type f -print -exec mv {} . \; && rm miopen-opencl_1.6.0-ffedda8_amd64.deb hip-thrust_1.8.2_all.deb && dpkg -i *.deb
sudo reboot
```
Please properly backup your system before proceeding, there's always risks to touch the kernel drivers :-)

**Comment by vidullan** [comment]
> Hi @vidullan , did the application work when you had ROCm1.9? This really sounds like a local environmental issue.
> You can use the following steps to clean your current ROCm packages:
> `sudo apt autoremove -y rocm_bandwidth_test rocminfo hsakmt-roct hsa-rocr-dev rocm-cmake hsa-ext-rocr-dev rocm-device-libs hip_base hip_samples rock-dkms`
> 
> Then you can download the ROCm1.9.2 archieve and install the deb packages, the following steps should work:
> 
> ```
> cd ~ && wget http://repo.radeon.com/rocm/archive/apt_1.9.2.tar.bz2 && tar -xvf apt_1.9.2.tar.bz2
> cd ~/apt_1.9.2.307 && cd pool/main && find . -mindepth 2 -type f -print -exec mv {} . \; && rm miopen-opencl_1.6.0-ffedda8_amd64.deb hip-thrust_1.8.2_all.deb && dpkg -i *.deb
> sudo reboot
> ```
> 
> Please properly backup your system before proceeding, there's always risks to touch the kernel drivers :-)

Yes, it was previously working which is why I quoted the time per step as 3.4 seconds per step while training on Vega64. I only realised something was wrong when I trained on my other system with a 1070 and got a per step time of 0.3.

Anyways, I rebuilt my whole system. Fresh install of ubuntu 18.04 along with the latest rocm and tensorflow installed as the readme suggested. Similar issue. Takes 353 seconds for the first step then  4 seconds for the subsequent steps. 

Unless the Vega64 is meant to be 10x slower than the 1070 surely there is an issue.

**Comment by vidullan** [comment]
Using TF_ROCM_FUSION_ENABLE= 1 leads to a core dump. 

2019-03-08 18:09:04.491579: I tensorflow/core/graph/gpu_fusion_pass.cc:434] ROCm Fusion is enabled.
2019-03-08 18:09:04.516371: F tensorflow/core/graph/gpu_fusion_pass.cc:1162] Non-OK-status: norm->input_node(2, &norm_offset) status: Invalid argument: Invalid input_edge index: 2, Node FirstStageFeatureExtractor/InceptionV2/Conv2d_1a_7x7/BatchNorm/moving_mean/read only has 1 inputs.
Aborted (core dumped)


Does that command change the computational graph?

**Comment by sunway513** [comment]
> Does that command change the computational graph?

Yes, by enabling Tensorflow fusion the computational graph would be changed. To further isolate your issue, could you try to follow the [doc](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/rocm_docs/rocm-port-overview.md#fusion-support) and try to disable our supported fusion ops one by one?
```
set TF_ROCM_FUSION_DISABLE_CBA to 1 to disable to Convolution+Bias+Activation fusion (forward and inference)
set TF_ROCM_FUSION_DISABLE_BNA to 1 to disable to BatchNorm+Activation fusions (forward, backward and inference)
set TF_ROCM_FUSION_DISABLE_ADDRELU to 1 to disable to Add+Relu fusion
set TF_ROCM_FUSION_DISABLE_ADDNRELUGRAD to 1 to disable to AddN+ReluGrad fusion
```

**Comment by ppanchad-amd** [comment]
Hi @vidullan. Closing ticket due to lack of activity and Vega64 is not supported on the latest ROCm. Please feel to open another ticket if you have any questions/issues with the latest ROCm.  Thanks!
